fdbserver/DataDistribution.actor.cpp

```diff
@@ -1442,165 +1453,290 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
-	audit->actors.add(loadAndDispatchAuditRange(self, audit, auditState.range));
+	audit->coreState.setPhase(AuditPhase::Running);
+	audit->retryCount = retryCount;
+	ASSERT(!self->audits[audit->coreState.getType()].contains(audit->coreState.id));
+	self->audits[audit->coreState.getType()][audit->coreState.id] = audit;
+	audit->actors.add(loadAndDispatchAuditRange(self, audit, audit->coreState.range));
 	TraceEvent(SevDebug, "DDResumeAuditStorageBegin", self->ddId)
-	    .detail("AuditID", audit->id)
-	    .detail("Range", auditState.range)
-	    .detail("AuditType", auditState.type);
-	state int retryTime = 0;
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("Range", audit->coreState.range)
+	    .detail("AuditType", audit->coreState.getType());
 	loop {
 		try {
 			wait(audit->actors.getResult());
 			TraceEvent(SevDebug, "DDResumeAuditStorageEnd", self->ddId)
-			    .detail("AuditID", audit->id)
-			    .detail("Range", auditState.range)
-			    .detail("AuditType", auditState.type);
-			auditState.setPhase(AuditPhase::Complete);
-			wait(persistAuditState(self->txnProcessor->context(), auditState));
+			    .detail("AuditID", audit->coreState.id)
+			    .detail("Range", audit->coreState.range)
+			    .detail("AuditType", audit->coreState.getType());
+			if (audit->foundError) {
+				audit->coreState.setPhase(AuditPhase::Error);
+			} else {
+				audit->coreState.setPhase(AuditPhase::Complete);
+			}
+			wait(persistAuditState(self->txnProcessor->context(), audit->coreState));
 		} catch (Error& e) {
 			TraceEvent(SevInfo, "DDResumeAuditStorageOperationError", self->ddId)
 			    .errorUnsuppressed(e)
-			    .detail("AuditID", audit->id)
-			    .detail("Range", auditState.range)
-			    .detail("AuditType", auditState.type);
-			if (e.code() == error_code_audit_storage_error) {
-				auditState.setPhase(AuditPhase::Error);
-				wait(persistAuditState(self->txnProcessor->context(), auditState));
-			} else if (retryTime > 10) {
-				auditState.setPhase(AuditPhase::Failed);
-				wait(persistAuditState(self->txnProcessor->context(), auditState));
-			} else if (e.code() != error_code_actor_cancelled) {
-				wait(delay(30));
-				retryTime++;
+			    .detail("AuditID", audit->coreState.id)
+			    .detail("Range", audit->coreState.range)
+			    .detail("AuditType", audit->coreState.getType())
+			    .detail("RetryCount", audit->retryCount);
+			if (e.code() == error_code_actor_cancelled) {
+				throw e;
+			} else if (audit->retryCount > 30) {
+				audit->coreState.setPhase(AuditPhase::Failed);
+				wait(persistAuditState(self->txnProcessor->context(), audit->coreState));
+			} else {
+				wait(delay(5));
+				audit->retryCount++;
 				continue;
 			}
 		}
 		break;
 	}
-	self->audits[auditState.getType()].erase(audit->id);
+	ASSERT(audit->coreState.getPhase() == AuditPhase::Failed || audit->coreState.getPhase() == AuditPhase::Error ||
+	       audit->coreState.getPhase() == AuditPhase::Complete);
+	audit->actors.clear(true);
```

Comment:
Do we need this? At this time, the actors have already completed or failed?

Reply:
Yes